### PR TITLE
Address performance issues with stencil and block-cyclic array init

### PR DESCRIPTION
**Pull request by mppf**
Follow-up to PR #15239.
Related to https://github.com/Cray/chapel-private/issues/992

The array initialization improvements (PR #15239) introduced performance
regressions for StencilDist and BlockCyclicDist arrays. Some of these
were due to increased communication. The reason that Stencil and
BlockCyclic arrays are affected is that they both allocate elements that
are not initialized by array initialization (in the case of initializing
from another same-distribution array, say). For Stencil these are
fluff/ghost cell elements. For BlockCyclic they are padding.

See the test performance/array/distCreate.chpl run for example with
    
```
$ ./distCreate --size=arraySize.small --dist=distType.blockCyc -nl 2
```

This PR resolves these problems by:
 * initializing the padding/fluff elements along with the original data
   when default initializing
 * initializing the padding/fluff elements always (because they won't be
   initialized by an initialization expression e.g.) and deinitializing
   them always (because they won't be moved out of when move-initializing
   arrays either)
 * using more efficient means to identify elements that are fluff/padding 

- [x] full local gasnet testing
- [x] full local testing

Reviewed by @e-kayrakli - thanks!

### CODE REVIEW

File: modules/dists/BlockCycDist.chpl [hunk 1]
```diff
@@ -1102,100 +1110,156 @@ class LocBlockCyclicArr {
   const numblocks : rank*int = allocDom._lens;
 
   const sizes : (rank+1)*int = allocDom._sizes;
+  const localeIndex: if rank == 1 then int else rank*int;
```
Comment (e-kayrakli) [inline modules/dists/BlockCycDist.chpl:1113]: I am a bit surprised this compiles. I would expect `type localeIndex`

File: modules/dists/BlockCycDist.chpl [hunk 2]
```diff
@@ -1102,100 +1110,156 @@ class LocBlockCyclicArr {
...
 
+    /*writeln(" whole= ", allocDom.globDom.whole,
+            " localeIndex= ", localeIndex,
+            " blocksize= ", blocksize,
+            " low= ", low,
+            " locsize= ", locsize,
+            " numblocks= ", numblocks,
+            " sizes= ", sizes,
+            " globalIndex= ", i,
+            " flatIndex= ", idx);*/
```
Comment (e-kayrakli) [inline modules/dists/BlockCycDist.chpl:1219]: Making sure that you want this comment to be merged in. We have some other examples of this AFAIK and I am OK with it as long as it is intentional

File: modules/dists/BlockCycDist.chpl [hunk 3]
```diff
@@ -886,9 +887,16 @@ override proc BlockCyclicArr.dsiDestroyArr(param deinitElts:bool) {
     on dom.dist.targetLocales(localeIdx) {
       var arr = locArr(localeIdx);
       if deinitElts {
-        for subdom in do_dsiLocalSubdomains(arr!.indexDom) {
-          forall j in subdom {
-            chpl__autoDestroy(arr(j));
+        param needsDestroy = __primitive("needs auto destroy", eltType);
```
Comment (e-kayrakli) [inline modules/dists/BlockCycDist.chpl:890]: Can this be a helper in Types.chpl?

File: modules/dists/BlockCycDist.chpl (hunk 1 above)
Comment (mppf) [reply modules/dists/BlockCycDist.chpl:1113]: It's declaring a variable, not a type; it's just doing so while specifying the type.

File: modules/dists/BlockCycDist.chpl (hunk 2 above)
Comment (mppf) [reply modules/dists/BlockCycDist.chpl:1219]: I removed it and rewrote it 3 times or so during debugging so wanted to leave it in until testing passed at least. I think I'll remove it before merging.

File: modules/dists/BlockCycDist.chpl (hunk 1 above)
Comment (e-kayrakli) [reply modules/dists/BlockCycDist.chpl:1113]: Ugh... sorry I got confused

**Pull request merged**